README
NAME
    Sys::VolumeGroups - Retrieve list of volume groups, logical volumes and
    their properties

SYNOPSIS
        my $sysvgs = Sys::VolumeGroups->new();
        my @vgs = $sysvgs->get_physical_volumes();
        print "Volume groups: ", join( ", ", map { $_->{VOLUME_GROUP} } @vgs ), "\n";
        my @lvs = $sysvgs->get_logical_volumes();
        print "Logical volumes: ", join( ", ", map { $_->{LOGICAL_VOLUME} } @lvs ), "\n";

SUBROUTINES/METHODS
  new
    Instantiates new Sys::VolumeGroups object. Attributes can be specified
    for used finder (of type Module::Pluggable::Object). Additionally,

    `only_loaded'
        Use only plugins which are still loaded.

    can be specified with a true value. This forces to grep `%INC' instead
    of using Module::Pluggable.

  get_physical_volumes
  get_logical_volumes
AUTHOR
    Jens Rehsack, `<rehsack at cpan.org>'

BUGS
    Please report any bugs or feature requests to `bug-sys-volumegroups at
    rt.cpan.org', or through the web interface at
    http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Sys-VolumeGroups. I will
    be notified, and then you'll automatically be notified of progress on
    your bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc Sys::VolumeGroups

    You can also look for information at:

    * RT: CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Sys-VolumeGroups

    * AnnoCPAN: Annotated CPAN documentation
        http://annocpan.org/dist/Sys-VolumeGroups

    * CPAN Ratings
        http://cpanratings.perl.org/d/Sys-VolumeGroups

    * Search CPAN
        http://search.cpan.org/dist/Sys-VolumeGroups/

ACKNOWLEDGEMENTS
LICENSE AND COPYRIGHT
    Copyright 2010 Jens Rehsack.

    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published
    by the Free Software Foundation; or the Artistic License.

    See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Sys-VolumeGroups

0.01    Date/time
        First version, released on an unsuspecting world.



lib/Sys/VolumeGroups/AixLvm.pm
package Sys::VolumeGroups::AixLvm;

use warnings;
use strict;

require 5.008;
require IPC::Cmd;

=head1 NAME

Sys::VolumeGroups::AixLvm - provides the access to lvm information on AIX

=cut

our $VERSION = '0.001';

require IPC::Cmd;
my ( $lsvg, $lslv );

sub usable
{
    unless( defined( $lsvg ) )
    {
	$lsvg = IPC::Cmd::can_run( 'lsvg' );
	$lsvg ||= '';
    }

    unless( defined( $lslv ) )
    {
	$lslv = IPC::Cmd::can_run( 'lslv' );
	$lslv ||= '';
    }

    return $lsvg && $lslv;
}

sub new
{
    my ($class, %opts) = $_[0];

    $opts{lsvg} ||= $lsvg;
    $opts{lslv} ||= $lslv;

    my $self = bless( \%opts, $class );

    return $self;
}

# see demo table at http://publib.boulder.ibm.com/infocenter/aix/v6r1/index.jsp?topic=/com.ibm.aix.cmds/doc/aixcmds3/lsvg.htm
sub _parse_table
{
    my $tblbuf = $_[0];
    my @lines = split( "\n", $tblbuf );
    my %table;

    foreach my $line (@lines)
    {
	while (
	    $tblbuf =~ m/
			\G(?:^|[\s]+|\b)        # start ...
			([^\s][^:]+):           # identifier
			\s*                     # separator
			(                       # group data
			    (?:
				[^\s:]+         # not a white space
				(?:             # and ...
				    \s[^\s:]+   # white space followed by an identifier without ':' at the end
				)*              # 0..n times
			    )+                  # at least once
			)
			(?:[\s]+?|$)            # followed by finishing white space or EOL
		    /gx
	      )
	{
	    my $descr = $1;
	    my $data = $2;
	    $descr =~ s/\W+$//;
	    $descr =~ s/\W+/_/g;
	    $table{ uc $descr } = $data;
	}
    }

    return \%table;
}

#Total PVs
#Active PVs
#VG identifier
#PP size
#Total PPs
#Free PPs
#Alloc PPs
#Quorum
#Auto-on
#Concurrent
#Auto-Concurrent
#VG Mode
#Node ID
#Active Nodes
#Max PPs Per PV
#Max PVs
#LTG size
#BB POLICY
#SNAPSHOT VG
#PRIMARY VG

my %vg_descriptor_map = (
    'VOLUME_GROUP_STATE' => 'VG_STATE',
    'PERMISSION' => 'VG_PERMISSION',
    'VGDS' => 'VG_DESCRIPTORS', # XXX ???
);

sub get_physical_volumes
{
    my $self = $_[0];
    my @vgs;
    my %online;

    my ( $success, $error_code, $full_buf, $stdout_buf, $stderr_buf ) =
      IPC::Cmd::run( command => [ $self->{lsvg}, '-L', '-o' ],
                     verbose => 0, );

    if ($success)
    {
	chomp $stdout_buf->[0];
	my @online_vgs = grep { m/^\w+$/ } split($stdout_buf->[0]);

	%online = map { $_ => 1 } @online_vgs;
	( $success, $error_code, $full_buf, $stdout_buf, $stderr_buf ) =
	  IPC::Cmd::run( command => [ $self->{lsvg}, '-L', @online_vgs ],
			 verbose => 0, );

	if ($success)
	{
	    chomp $stdout_buf->[0];
	    # split from 'Volume group.*' -> 'Volume group' or end
	    my $vgdatatbl = _parse_table( $stdout_buf->[0] );
	    my %vgdata = map { ( $vg_descriptor_map{$_} || $_ ) => $vgdatatbl->{$_} } keys %$vgdatatbl;
	    push( @vgs, \%vgdata );
	}

	( $success, $error_code, $full_buf, $stdout_buf, $stderr_buf ) =
	  IPC::Cmd::run( command => [ $self->{lsvg}, '-L' ],
			 verbose => 0, );

	if ($success)
	{
	    chomp $stdout_buf->[0];
	    my @all_vgs = grep { m/^\w+$/ } split($stdout_buf->[0]);

	    foreach my $vg (@all_vgs)
	    {
		next if( $online{$vg};
		my %vgdata = (
		    VOLUME_GROUP => $vg,
		    VG_STATE => 'offline',
		);
		push( @vgs, \%vgdata );
	    }
	}
    }

    return @vgs;
}

sub get_logical_volumes
{
    my $self = $_[0];
    my @vgs;

    my ( $success, $error_code, $full_buf, $stdout_buf, $stderr_buf ) =
      IPC::Cmd::run( command => [ $self->{lsvg}, '-L', '-o' ],
                     verbose => 0, );

    if ($success)
    {
	chomp $stdout_buf->[0];
	my @online_vgs = grep { m/^\w+$/ } split($stdout_buf->[0]);

	%online = map { $_ => 1 } @online_vgs;
	( $success, $error_code, $full_buf, $stdout_buf, $stderr_buf ) =
	  IPC::Cmd::run( command => [ $self->{lsvg}, '-L', '-l', @online_vgs ],
			 verbose => 0, );

	if ($success)
	{
	    chomp $stdout_buf->[0];
	    my $vgdatatbl = _parse_table( $stdout_buf->[0] );
	    my %vgdata = map { ( $vg_descriptor_map{$_} || $_ ) => $vgdatatbl->{$_} } keys %$vgdatatbl;
	    push( @vgs, \%vgdata );
	}

    }

    return @vgs;
}

=head1 AUTHOR

Jens Rehsack, C<< <rehsack at cpan.org> >>

=head1 LICENSE AND COPYRIGHT

Copyright 2010 Jens Rehsack.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1;



lib/Sys/VolumeGroups/LinuxLvm2.pm
package Sys::VolumeGroups::LinuxLvm2;

use warnings;
use strict;

require 5.008;
require IPC::Cmd;

=head1 NAME

Sys::VolumeGroups::LinuxLvm2 - provides the access to lvm2 information

=cut

our $VERSION = '0.001';

require IPC::Cmd;
use List::Moreutils qw(zip);

my ( $vgdisplay, $lvdisplay);

sub usable
{
    unless( defined( $vgdisplay ) )
    {
	$vgdisplay = IPC::Cmd::can_run( 'vgdisplay' );
	$vgdisplay ||= '';
    }

    unless( defined( $lvdisplay ) )
    {
	$lvdisplay = IPC::Cmd::can_run( 'lvdisplay' );
	$lvdisplay ||= '';
    }

    return $vgdisplay && $lvdisplay;
}

sub new
{
    my ($class, %opts) = $_[0];

    $opts{vgdisplay} ||= $vgdisplay;
    $opts{lvdisplay} ||= $lvdisplay;

    my $self = bless( \%opts, $class );

    return $self;
}

sub _trim
{
    my $s = $[0];
    $s =~ s/^\s*//g;
    $s =~ s/\s*$//g;
    return $s;
}

my @vg_descriptors = qw(VOLUME_GROUP VG_PERMISSION VG_STATE VG_DESCRIPTORS
      MAX_LVS LVS OPEN_LVS VG_LG_MAXSIZE MAX_PVS TOTAL_PVS
      ACTIVE_PVS VG_SIZE PP_SIZE TOTAL_PPS
      USED_PPS FREE_PPS VG_IDENTIFIER);

sub get_physical_volumes
{
    my $self = $_[0];
    my @vgs;

    my ( $success, $error_code, $full_buf, $stdout_buf, $stderr_buf ) =
      IPC::Cmd::run( command => [ $self->{vgdisplay} ],
                     verbose => 0, );

    if ($success)
    {
	chomp $stdout_buf->[0];
	my @outbuf = split($stdout_buf->[0]);
        foreach my $line ( @outbuf )
        {
            $line = _trim($line);

            my @data = split( m/:/, $line, 17 );
	    next if( scalar @data < 17 );
	    my %vgdata = zip( @vg_descriptors, @data );
            push( @vgs, \%vgdata );
        }
    }

    return @vgs;
}

my @lv_descriptors = qw(LOGICAL_VOLUME VOLUME_GROUP LV_PERMISSION LV_STATE LV_NUMBER
      LV_COUNT LV_SIZE LV_EXT_AMOUNT LV_EXT_ALLOC LV_POLICY
      LV_READAHEAD LV_MAJOR LV_MINOR LV_MAPPED_PATH);

sub get_logical_volumes
{
    my $self = $_[0];
    my @lvs;

    my ( $success, $error_code, $full_buf, $stdout_buf, $stderr_buf ) =
      IPC::Cmd::run( command => [ $self->{lvdisplay} ],
                     verbose => 0, );

    if ($success)
    {
	chomp $stdout_buf->[0];
	my @outbuf = split($stdout_buf->[0]);
        foreach my $line ( @outbuf )
        {
            $line = _trim($line);

            my @data = split( m/:/, $line, 13 );
	    next if( scalar @data < 13 );

            if ( -r $data[0] )
            {
                $data[13] = readlink( $data[0] );
            }
            else
            {
                $data[13] = undef;
            }
	    my %lvdata = zip( @lv_descriptors, @data );
            push( @lvs, \%lvdata );
        }
    }

    return @lvs;
}

=head1 AUTHOR

Jens Rehsack, C<< <rehsack at cpan.org> >>

=head1 LICENSE AND COPYRIGHT

Copyright 2010 Jens Rehsack.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1;


lib/Sys/VolumeGroups.pm
package Sys::VolumeGroups;

use warnings;
use strict;

require 5.008;
require Module::Pluggable::Object;

=head1 NAME

Sys::VolumeGroups - Retrieve list of volume groups, logical volumes and their properties

=cut

our $VERSION = '0.001';


=head1 SYNOPSIS

    my $sysvgs = Sys::VolumeGroups->new();
    my @vgs = $sysvgs->get_physical_volumes();
    print "Volume groups: ", join( ", ", map { $_->{VOLUME_GROUP} } @vgs ), "\n";
    my @lvs = $sysvgs->get_logical_volumes();
    print "Logical volumes: ", join( ", ", map { $_->{LOGICAL_VOLUME} } @lvs ), "\n";

=head1 SUBROUTINES/METHODS

=head2 new

Instantiates new Sys::VolumeGroups object. Attributes can be specified
for used finder (of type L<Module::Pluggable::Object>). Additionally,

=over 4

=item C<only_loaded>

Use only plugins which are still loaded.

=back

can be specified with a true value. This forces to grep C<%INC> instead
of using Module::Pluggable.

=cut

sub new
{
    my ( $class, %attrs ) = @_;
    my $self = bless( { plugins => [], }, $class );

    my $only_loaded = delete $attrs{only_loaded};

    if ($only_loaded)
    {
        my @search_path = __PACKAGE__ eq $class ? (__PACKAGE__) : ( __PACKAGE__, $class );
        foreach my $path (@search_path)
        {
            $path =~ s|::|/|g;
            $path .= "/";
            my @loadedModules = grep { 0 == index( $_, $path ) } keys %INC;
            foreach my $module (@loadedModules)
            {
                $module =~ s|/|::|;
                $module =~ s/\.pm$//;
                next unless ( $module->can('usable') && $module->usable() );
                push( @{ $self->{plugins} }, $module->new() );
            }
        }
    }
    else
    {
        %attrs = (
                   require     => 1,
                   search_path => [ __PACKAGE__ eq $class ? __PACKAGE__ : ( __PACKAGE__, $class ) ],
                   inner       => 0,
                   %attrs,
                 );
        my $finder  = Module::Pluggable::Object->new(%attrs);
        my @plugins = $finder->plugins();
        foreach my $plugin (@plugins)
        {
            next unless ( $plugin->can('usable') && $plugin->usable() );
            push( @{ $self->{plugins} }, $plugin->new() );
        }
    }

    return $self;
}

=head2 get_physical_volumes

=cut

sub get_physical_volumes
{
    my $self = $_[0];
    my @vgs;

    foreach my $plugin (@{$self->{plugins}})
    {
	push( @vgs, $plugin->get_physical_volumes() );
    }

    return @vgs;
}

=head2 get_logical_volumes

=cut

sub get_logical_volumes
{
    my $self = $_[0];
    my @lvs;

    foreach my $plugin (@{$self->{plugins}})
    {
	push( @lvs, $plugin->get_logical_volumes() );
    }

    return @lvs;
}

=head1 AUTHOR

Jens Rehsack, C<< <rehsack at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to
C<bug-sys-volumegroups at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Sys-VolumeGroups>.  I
will be notified, and then you'll automatically be notified of progress
on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Sys::VolumeGroups

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Sys-VolumeGroups>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Sys-VolumeGroups>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Sys-VolumeGroups>

=item * Search CPAN

L<http://search.cpan.org/dist/Sys-VolumeGroups/>

=back

=head1 RESOURCES AND CONTRIBUTIONS

There're several ways how you can help to support future development: You
can hire the author to implement the features you require at most (this
also defines priorities), you can negotiate a support and maintenance
contract with the company of the author and you can provide tests and
patches. Further, you can submit documentation and links to resources to
improve or add volume group managers or grant remote access to machines
with insufficient supported volume managers, file systems with volume
manager capabilities or storage systems.

=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2010 Jens Rehsack.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1; # End of Sys::VolumeGroups
